### PR TITLE
Fix races related to OS scheduling and spurious wakeup

### PR DESCRIPTION
**Pull request by wks**
This PR fixes two race conditions.

1.  If a GC worker spuriously wakes up after the `Finish` coordinator message is sent, the coordinator may deactivate work packets while the GC worker is trying to schedule sentinels or open new buckets.

2.  If `Plan::schedule_collection` runs too slowly due to OS scheduling, workers may open subsequent work buckets while `Plan::schedule_collection` still has pending packets to be added in prior buckets.

Fixes: https://github.com/mmtk/mmtk-core/issues/758

### CODE REVIEW

File: src/scheduler/controller.rs [hunk 1]
```diff
@@ -128,7 +143,11 @@ impl<VM: VMBinding> GCController<VM> {
                 CoordinatorMessage::Finish => {}
             }
         }
-        self.scheduler.deactivate_all();
+
+        {
+            let _guard = self.scheduler.worker_monitor.0.lock().unwrap();
```
Comment (qinsoon) [inline src/scheduler/controller.rs:148]: Maybe add some comments about why we need to acquire the lock here (like 'make sure no worker is trying to update buckets or schedule sentinel while we are deactivating buckets').

**Review by qinsoon — APPROVED** [review]
LGTM

File: src/scheduler/controller.rs [hunk 2]
```diff
@@ -101,7 +101,22 @@ impl<VM: VMBinding> GCController<VM> {
     /// Coordinate workers to perform GC in response to a GC request.
     pub fn do_gc_until_completion(&mut self) {
         // Schedule collection.
+        // Note: GC workers will start executing work packets as soon as individual work packets
+        // are added.  This means workers may open subsequent buckets while `schedule_collection`
+        // still has packets to be added to prior buckets.
+        // We use the`pending_coordinator_packets` to prevent the workers from opening any work
+        // buckets.
+        self.scheduler
```
[the remark this answers is not in the document]
Comment (wks) [reply src/scheduler/controller.rs]: I have changed the code accordingly.  I added an `initiate_coordinator_work` function for the controller itself to start a coordinator work packet, and an `execute_coordinator_work` function to execute coordinator work packets sent from workers.

File: src/scheduler/controller.rs (hunk 1 above)
Comment (wks) [reply src/scheduler/controller.rs:123]: I added a comment.

**Pull request merged**